### ISSUE TITLE
Mod options - further work

### ISSUE THREAD

**Issue by aubergine10** [CitiesSkylinesMods/TMPE#1452]
Additional stuff part of #1356

- Find better way to do default values of mod options
    - Notably for use in Scenario Editor (from map) where `OptionsManager.LoadData()` isn't getting called
- Display error messages when load/save fails
- Don't display/log "no data" errors when starting new game from map/scenario

**Comment by krzychu124** [comment]
> Notably for use in Scenario Editor (from map) where OptionsManager.LoadData() isn't getting called

That's not true.

<details>

```cs
...
Info 208.3676596: Loading Traffic Manager: PE Data
Info 208.3687276: Initializing flags
Info 208.3821517: OnBeforeLoadData: ExtNodeManager
Info 208.3831633: OnBeforeLoadData: ExtSegmentManager
Debug 208.3856770: ExtSegmentManager.OnBeforeLoadData: Calculating 36864 extended segments...
Debug 208.6670544: ExtSegmentManager.OnBeforeLoadData: Calculation finished.
Info 208.6679269: OnBeforeLoadData: ExtSegmentEndManager
Info 208.6687831: OnBeforeLoadData: GeometryManager
Info 208.7158434: OnBeforeLoadData: AdvancedParkingManager
Info 208.7168257: OnBeforeLoadData: CustomSegmentLightsManager
Info 208.7176791: OnBeforeLoadData: ExtBuildingManager
Info 208.7187096: OnBeforeLoadData: ExtCitizenInstanceManager
Info 208.7195425: OnBeforeLoadData: ExtCitizenManager
Info 208.7203620: OnBeforeLoadData: ExtPathManager
Info 208.7211683: OnBeforeLoadData: TurnOnRedManager
Info 208.7395141: OnBeforeLoadData: LaneArrowManager
Info 208.7404819: OnBeforeLoadData: LaneConnectionManager
Info 208.7413101: OnBeforeLoadData: OptionsManager
Info 208.7425367: OnBeforeLoadData: ParkingRestrictionsManager
Info 208.7517249: OnBeforeLoadData: RoutingManager
Info 208.7526106: OnBeforeLoadData: SpeedLimitManager
Info 208.7642350: OnBeforeLoadData: TrafficLightManager
Info 208.7651650: OnBeforeLoadData: TrafficLightSimulationManager
Info 208.7660197: OnBeforeLoadData: TrafficMeasurementManager
Info 208.7668623: OnBeforeLoadData: TrafficPriorityManager
Info 208.7676919: OnBeforeLoadData: UtilityManager
Info 208.7685064: OnBeforeLoadData: VehicleRestrictionsManager
Info 208.7693455: OnBeforeLoadData: ExtVehicleManager
Info 208.7701625: OnBeforeLoadData: RoadSignThemes
Info 208.7709777: OnBeforeLoadData: JunctionRestrictions
Info 208.7717855: OnBeforeLoadData: RoadUI
Info 208.7725958: OnBeforeLoadData: TrafficLightTextures
Info 208.7734255: OnBeforeLoadData: JunctionRestrictionsManager
Info 208.7745675: Initialization done. Loading mod data now.
Info 208.7765259: No VersionInfo data to deserialize!
Info 208.7778387: Reading VersionInfo from config
Warning 208.7791877: Version configuration NULL, Couldn't load data. Possibly a new game?
   at CSUtil.Commons.Log.LogToFile(System.String log, LogLevel level)
   at CSUtil.Commons.Log.Warning(System.String s)
   at TrafficManager.Lifecycle.SerializableDataExtension.ReportVersionInfo(Boolean ByRef error)
   at TrafficManager.Lifecycle.SerializableDataExtension.DeserializeVersionData(System.Byte[] data)
   at TrafficManager.Lifecycle.SerializableDataExtension.Load()
   at TrafficManager.Lifecycle.SerializableDataExtension.OnLoadData()
   at SerializableDataWrapper.OnLoadData()
   at SimulationManager.LateUpdateData(UpdateMode mode, Single minProgress, Single maxProgress)
   at SimulationManager.Managers_LateUpdateData(UpdateMode mode, Single minProgress, Single maxProgress)
   at LoadingManager+<LoadSimulationData>c__IteratorB.MoveNext()
   at AsyncTask.Execute()
   at SimulationManager.SimulationStep()
   at SimulationManager.SimulationThread()

Info 208.8250960: No data to deserialize!
Info 208.8275895: Loading State from Config
Warning 208.8284168: Configuration NULL, Couldn't load save data. Possibly a new game?
   at CSUtil.Commons.Log.LogToFile(System.String log, LogLevel level)
   at CSUtil.Commons.Log.Warning(System.String s)
   at TrafficManager.Lifecycle.SerializableDataExtension.LoadDataState(Boolean ByRef error)
   at TrafficManager.Lifecycle.SerializableDataExtension.DeserializeData(System.Byte[] data)
   at TrafficManager.Lifecycle.SerializableDataExtension.Load()
   at TrafficManager.Lifecycle.SerializableDataExtension.OnLoadData()
   at SerializableDataWrapper.OnLoadData()
   at SimulationManager.LateUpdateData(UpdateMode mode, Single minProgress, Single maxProgress)
   at SimulationManager.Managers_LateUpdateData(UpdateMode mode, Single minProgress, Single maxProgress)
   at LoadingManager+<LoadSimulationData>c__IteratorB.MoveNext()
   at AsyncTask.Execute()
   at SimulationManager.SimulationStep()
   at SimulationManager.SimulationThread()

Info 208.8295428: OnLoadData completed successfully.
Info 208.8305016: OnAfterLoadData: ExtNodeManager
Info 208.8314993: OnAfterLoadData: ExtSegmentManager
Info 208.8323161: OnAfterLoadData: ExtSegmentEndManager
Info 208.8331526: OnAfterLoadData: GeometryManager
Info 208.8339807: OnAfterLoadData: AdvancedParkingManager
Info 208.8348438: OnAfterLoadData: CustomSegmentLightsManager
Info 208.8360743: OnAfterLoadData: ExtBuildingManager
Info 208.8376960: OnAfterLoadData: ExtCitizenInstanceManager
Info 208.8388226: OnAfterLoadData: ExtCitizenManager
Info 208.8397226: OnAfterLoadData: ExtPathManager
Info 208.8405700: OnAfterLoadData: TurnOnRedManager
Info 208.8413997: OnAfterLoadData: LaneArrowManager
Info 208.9061988: OnAfterLoadData: LaneConnectionManager
Info 208.9075772: OnAfterLoadData: OptionsManager
Info 208.9093932: OptionsManger.OnAfterLoadData() checking for queued method calls
Info 208.9102176: OnAfterLoadData: ParkingRestrictionsManager
Info 208.9110557: OnAfterLoadData: RoutingManager
Debug 208.9126393: RoutingManager.RecalculateAll: called
Info 209.2525109: OnAfterLoadData: SpeedLimitManager
Info 209.2539289: OnAfterLoadData: TrafficLightManager
Info 209.2552661: OnAfterLoadData: TrafficLightSimulationManager
Info 209.2566497: OnAfterLoadData: TrafficMeasurementManager
Info 209.2578301: OnAfterLoadData: TrafficPriorityManager
Info 209.2587815: OnAfterLoadData: UtilityManager
Info 209.2596805: OnAfterLoadData: VehicleRestrictionsManager
Info 209.2606671: OnAfterLoadData: ExtVehicleManager
Debug 209.2637010: ExtVehicleManager: InitAllVehicles()
Info 209.2687873: OnAfterLoadData: RoadSignThemes
Info 209.2699977: OnAfterLoadData: JunctionRestrictions
Info 209.2710406: OnAfterLoadData: RoadUI
Info 209.2720156: OnAfterLoadData: TrafficLightTextures
Info 209.2731032: OnAfterLoadData: JunctionRestrictionsManager
Info 223.0611516: TMPELifecycle.Load() called. Mode=NewScenarioFromMap, UpdateMode=NewScenarioFromMap, Scene=ScenarioEditor
Info 223.0631260: Expected C:SL v1.14.0-f9 - Actual C:SL v1.14.0-f9
Debug 223.0643074: ModUI.OnLevelLoaded: called
Debug 223.0652236: Adding UIBase instance.
Debug 223.0706316: ##### Initializing ModUI.
Debug 223.8917524: StartPathFindCommons.ReplaceMaxPosTranspiler() successfully replaced 4 instances of ldc.r4 4800 with 8540
Debug 224.4546525: StartPathFindCommons.ReplaceMaxPosTranspiler() successfully replaced 4 instances of ldc.r4 4800 with 8540
Info 224.9425945: TMPE patches installed successfully
Info 224.9439215: Notifying managers...
Info 224.9449799: OnLevelLoading: ExtNodeManager
Info 224.9460719: OnLevelLoading: ExtSegmentManager
Info 224.9469227: OnLevelLoading: ExtSegmentEndManager
Debug 224.9482755: ExtSegmentEndManager.OnLevelLoading: Calculating 73728 extended segment ends...
Debug 225.0242185: ExtSegmentEndManager.OnLevelLoading: Calculation finished.
Info 225.0253427: OnLevelLoading: GeometryManager
Info 225.0263753: OnLevelLoading: AdvancedParkingManager
Info 225.0275609: OnLevelLoading: CustomSegmentLightsManager
Info 225.0295843: OnLevelLoading: ExtBuildingManager
Info 225.0305872: OnLevelLoading: ExtCitizenInstanceManager
Info 225.0314295: OnLevelLoading: ExtCitizenManager
Info 225.0322333: OnLevelLoading: ExtPathManager
Info 225.0330765: OnLevelLoading: TurnOnRedManager
Info 225.0338866: OnLevelLoading: LaneArrowManager
Info 225.0350633: OnLevelLoading: LaneConnectionManager
Info 225.0359059: OnLevelLoading: OptionsManager
Info 225.0367947: OnLevelLoading: ParkingRestrictionsManager
Info 225.0376255: OnLevelLoading: RoutingManager
Info 225.0384262: OnLevelLoading: SpeedLimitManager
Info 225.0392262: OnLevelLoading: TrafficLightManager
Info 225.0400216: OnLevelLoading: TrafficLightSimulationManager
Info 225.0408353: OnLevelLoading: TrafficMeasurementManager
Info 225.0416328: OnLevelLoading: TrafficPriorityManager
Info 225.0424935: OnLevelLoading: UtilityManager
Info 225.0432826: OnLevelLoading: VehicleRestrictionsManager
Info 225.0441007: OnLevelLoading: ExtVehicleManager
Info 225.0449107: OnLevelLoading: RoadSignThemes
Debug 225.1039914: Resource SignThemes.RoadDefaults.RightOfWay.png not found (not an error)
Debug 225.1051389: Resource SignThemes.RoadDefaults.Yield.png not found (not an error)
Debug 225.1060047: Resource SignThemes.RoadDefaults.Stop.png not found (not an error)
Debug 225.1068904: Resource SignThemes.RoadDefaults.Parking.png not found (not an error)
Debug 225.1082321: Resource SignThemes.RoadDefaults.NoParking.png not found (not an error)
Info 225.1132930: OnLevelLoading: JunctionRestrictions
Info 225.3861951: OnLevelLoading: RoadUI
Debug 225.4044118: Translated file RoadUI.sign_stop.png found
Debug 225.4111153: Translated file RoadUI.sign_yield.png found
Info 226.1195736: OnLevelLoading: TrafficLightTextures
Debug 226.2270112: Translated file TrafficLights.light_mode.png found
Debug 226.2288803: Translated file TrafficLights.light_counter.png found
Debug 226.2383845: Translated file TrafficLights.pedestrian_mode_1.png found
Debug 226.2402825: Translated file TrafficLights.pedestrian_mode_2.png found
Info 226.2793828: OnLevelLoading: JunctionRestrictionsManager
Info 226.3350979: OnLevelLoaded complete.
Debug 226.7773183: ModUI.EnableTool: called
Info 226.7785598: Initializing traffic manager tool...
Debug 226.7899965: TrafficManagerTool: Awake -504135680
Info 226.8006227: TrafficManagerTool: Initialization running now.
Debug 226.8205511: calling AddGuide(localeKey=LaneArrowTool_Disabled due to highway rules) ...
Debug 226.8222232: calling AddGuide(localeKey=LaneArrowTool_Disabled due to lane connections) ...
Debug 226.8231847: calling AddGuide(localeKey=TimedTrafficLightsTool_Auto TL no need) ...
Debug 226.8240537: calling AddGuide(localeKey=LaneConnectorTool:stay-in-lane is not supported for this setup) ...
Info 226.9232989: TrafficManagerTool.InitializeSubTools()
Debug 226.9464619: SetToolMode: not changed old=None new=None
Info 226.9878630: TrafficManagerTool: Initialization completed.
...
```

</details>

**Comment by aubergine10** [comment]
It is true :) First thing `OptionsManager.LoadData()` does is `Log.Info($"OptionsManager.LoadData: {data.Length} bytes");`

![image](https://user-images.githubusercontent.com/1386719/156476617-9683ca97-4017-452f-b3e8-d38a1944112f.png)

That's not appearing in log file. The `SerializeDataExtension` currently skips calling it because the data is `null` (because creating new scenario from map):

```csharp
                byte[] options = SerializableData.LoadData("TMPE_Options");
                if (options != null) {
                    if (!OptionsManager.Instance.LoadData(options)) {
                        loadingSucceeded = false;
                    }
                }
```

That's fixed by https://github.com/CitiesSkylinesMods/TMPE/pull/1425 which changes it to:

![image](https://user-images.githubusercontent.com/1386719/156477010-f51e9665-2d7d-4293-a269-fe50c706965c.png)
 :)

**Comment by krzychu124** [comment]
Anyways, I'm pretty sure you broke it some time ago, since it was working normally before all that "refactor movements"😂

**Comment by aubergine10** [comment]
Yup, I changed how defaults work - rather than being defined in `Options.cs` (which causes a bunch of different problems) the defaults are now set by `OptionsManager.LoadData()` ... but that won't get called in editors etc due to `null` data until #1425 is merged.

I think there is maybe some additional refinement for #1425 come to think of it... notably `ForceResetToDefaults` should return false in scneario and map editors (defaults will still occur if no data, but if data is present it should be loaded in those two editors specifically). I'll do that now and update that PR.

**Comment by aubergine10** [comment]
I decided not to load mod options in map/scenario editors, at least for now. We're not saving mod options in any of the editors (once #1425 is merged) so no point trying to load something that won't be there. Instead, default mod options (all mod features enabled) will be used.

Additionally, updated TTL mod option to allow TTL use in map/scenario editors. :) Will need some testing, obviously - but the tool now shows and can be used with the updated code from #1425 .

Been looking at the errors when stuff fails to load and it will require some refactoring as current approach is somewhat blunt in handling new game or new asset/map/scenario/etc (it thinks deserialization failed when in fact there was nothing to deserialize in those situations). Will work on separate PR to fix that stuff.
